### PR TITLE
Allow extensions in Add Products task

### PR DESCRIPTION
**Pull request by dakota**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

As part of the project to allow canonical WooCommerce extensions to surface themselves as part of the WooCommerce onboarding tasklist. This PR adds a filter to allow for such extensions to add themselves to the task-list.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install this WooCommerce Gift-Cards pre-release version
2. On the Product task list (`/wp-admin/admin.php?page=wc-admin&task=products`) you should see "Gift Card" as an option at the end of the list (hidden under the "view more" link)
3. Clicking it should create a new gift card product (This step isn't directly being tested as the primary work involved is in the gift cards plugin)

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message 

Add support for extending product types on onboarding task list

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-admin/client/task-lists/fills/products/dummyplugin.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/**
```
Comment (xristos3490) [inline plugins/woocommerce-admin/client/task-lists/fills/products/dummyplugin.js]: Is this file needed?

**Review by xristos3490 — COMMENTED** [review]
Looks good to me, @dakota! Let's polish it so @ilyasfoo can take a final look as well before I approve. Also left a minor comment.

The only thing that stands out to me is that this filter will allow complete replacements of that list; which can be a nice-to-have feature for heavily customized stores.

File: plugins/woocommerce-admin/client/task-lists/fills/products/dummyplugin.js (hunk 1 above)
Comment (dakota) [reply plugins/woocommerce-admin/client/task-lists/fills/products/dummyplugin.js]: It is not. It was created by @ilyasfoo as part of the initial POC code. I thought of leaving it in as a example, but can remove it.

Comment (ilyasfoo) [reply plugins/woocommerce-admin/client/task-lists/fills/products/dummyplugin.js]: Hey, @dakota! I'd recommend removing this file for review.

For a more realistic testing, appreciate if you could provide the GC plugin build that utilizes the filter. Thanks!

**Comment by dakota** [comment]
> The only thing that stands out to me is that this filter will allow complete replacements of that list; which can be a nice-to-have feature for heavily customized stores.

@xristos3490 - I've made a change to prevent replacing the list as per the [project requirements](https://woostartp2.wordpress.com/2024/02/13/pt-optimize-onboarding-of-product-type-feature-extensions/#project-plan).

**Comment by dakota** [comment]
@ilyasfoo @xristos3490 I believe this PR is now ready for CR and testing

File: plugins/woocommerce-admin/client/task-lists/fills/products/constants.tsx [hunk 2]
```diff
@@ -90,3 +90,6 @@ export const defaultSurfacedProductTypes =
 	onboardingProductTypesToSurfaced.physical;
 
 export const supportedOnboardingProductTypes = [ 'physical', 'downloads' ];
+
+export const SETUP_TASKLIST_PRODUCT_TYPES =
```
Comment (ilyasfoo) [inline plugins/woocommerce-admin/client/task-lists/fills/products/constants.tsx]: Constant name could be a tad clearer. What do you think?
```suggestion
export const SETUP_TASKLIST_PRODUCT_TYPES_FILTER =
```

File: plugins/woocommerce-admin/client/task-lists/fills/products/constants.tsx [hunk 3]
```diff
@@ -90,3 +90,6 @@ export const defaultSurfacedProductTypes =
...
+	'woocommerce_tasklist_experimental_product_types';
```
Comment (ilyasfoo) [inline plugins/woocommerce-admin/client/task-lists/fills/products/constants.tsx]: I looked at existing [experimental filters](https://github.com/woocommerce/woocommerce/blob/2f77de3f185644c0ef3b9b9cb5cdae6f9ca7b31e/plugins/woocommerce-admin/client/analytics/report/variations/table.js#L297), I think this is a better name
```suggestion
	'experimental_woocommerce_tasklist_product_types';
```

File: plugins/woocommerce-admin/client/task-lists/fills/products/utils.tsx [hunk 4]
```diff
@@ -2,31 +2,47 @@
...
+	SETUP_TASKLIST_PRODUCT_TYPES,
 } from './constants';
 
 export const getProductTypes = ( {
 	exclude,
 }: {
 	exclude?: ProductTypeKey[];
 } = {} ): ProductType[] => {
+	/**
+	 * Filter for adding custom product types to tasklist.
```
Comment (ilyasfoo) [inline plugins/woocommerce-admin/client/task-lists/fills/products/utils.tsx]: Similar suggestion with [existing experimental filter](https://github.com/woocommerce/woocommerce/blob/2f77de3f185644c0ef3b9b9cb5cdae6f9ca7b31e/plugins/woocommerce-admin/client/analytics/report/variations/table.js#L295)

```suggestion
	 * Experimental: Filter for adding custom product types to tasklist.
```

File: plugins/woocommerce-admin/client/task-lists/fills/products/utils.tsx [hunk 5]
```diff
@@ -2,31 +2,47 @@
...
+	 *
+	 * @filter woocommerce_tasklist_experimental_product_types
+	 * @param {Object} productTypes Array of product types.
+	 */
+	let productTypes = [
+		...baseProductTypes,
+		...( applyFilters(
```
Comment (ilyasfoo) [inline plugins/woocommerce-admin/client/task-lists/fills/products/utils.tsx]: I think we should make the extension point safer. 

AFAIK spreading (`...`) a non-iteratable object may result in an error. Let's check if the filter value is iteratable first.

**Review by ilyasfoo — CHANGES_REQUESTED** [review]
Testing well! However, I have few comments:

**Review by ilyasfoo — APPROVED** [review]
@dakota LGTM but please remember to update the filter name in GC plugin!

**Pull request merged**